bokehjs/src/coffee/models/sources/column_data_source.coffee

```diff
@@ -49,13 +50,58 @@ export stream_to_column = (col, new_col, rollover) ->
   return concat_typed_arrays(col, tmp)
 
 # exported for testing
-export patch_to_column = (col, patch) ->
-  for i in [0...patch.length]
-    [ind, value] = patch[i]
-    col[ind] = value
+export slice = (ind, length) ->
+  if isObject(ind)
+    return [ind.start ? 0, ind.stop ? length, ind.step ? 1]
+  return [start, stop, step] = [ind, ind+1, 1]
 
+# exported for testing
+export patch_to_column = (col, patch, shapes) ->
+  patched = new Set()
+  patched_range = false
+
+  for [ind, value] in patch
+
+    # make the single index case look like the length-3 multi-index case
+    if not isArray(ind)
 
-# Datasource where the data is defined column-wise, i.e. each key in the
+      if Number.isInteger(ind)
```

Comment:
> it really should be an integer, as it is an array index

Is there a distinction between and integers and floats in this code? If not (hopefully), then you shouldn't be validating here, just using minimal logic to distinguish javascript types. As to distinguishing non objects, this is tricky and you should always check for as specific types as possible, because otherwise:
```
> let t = require("./types")
undefined
> t.isObject(1)
false
> t.isObject(new Number(1))
true
```
Though I'm not sure if any code in bokehjs or any libraries we use produce this kind of objects.